packages/utils/telemetry-utils/src/errorLogging.ts

```diff
@@ -3,102 +3,203 @@
+    const props = isILoggingError(error) ? error.getTelemetryProperties() : {};
 
-    const message = (typeof error?.message === "string")
-        ? error.message as string
-        : String(error);
+    const newError = newErrorFn(message);
+    mixinTelemetryProps(newError, props);
 
-    const safeProps: { message: string; errorType?: string; stack?: string } = {
-        message,
-    };
+    if (stack !== undefined) {
+        Object.assign(newError, { stack });
+    }
 
-    if (isRegularObject(error)) {
-        const { errorType, stack, name } = error;
+    return newError;
+}
 
-        if (typeof errorType === "string") {
-            safeProps.errorType = errorType;
-        }
+/**
+ * Normalize the given error yielding a valid Fluid Error
+ * @returns The same error object passed in if possible, normalized and with any provided annotations applied
+ * @param error - The error to normalize, ideally by patching in the properties of IFluidErrorBase
+ * @param annotations - Annotations to apply to the normalized error
+ */
+export function normalizeError(
+    error: unknown,
+    annotations: FluidErrorAnnotations = {},
+): IFluidErrorBase {
+    if (isRegularObject(error)) {
+        assert(!Object.isFrozen(error), "Cannot normalize a frozen error object");
 
-        if (typeof stack === "string") {
-            const errorName = (typeof name === "string") ? name : undefined;
-            safeProps.stack = removeMessageFromStack(stack, errorName);
-        }
+        // We can simply annotate the error and return it
+        annotateErrorObject(error, annotations);
+        return error;
     }
 
-    return safeProps;
+    // We can't annotate the error, so we create a new one using its toString result as the message
+    const errorObject = new LoggingError(String(error));
+    const errorTypeIfNone = typeof(error);
+    const fluidError = patchFluidErrorBuilder(
+        errorObject as FluidErrorBuilder,
+        prepareFluidErrorTemplate(errorObject, errorTypeIfNone, annotations.errorCodeIfNone),
+    );
+
+    mixinTelemetryPropsWithFluidError(fluidError, annotations.props);
+    return fluidError;
 }
 
-/** Type guard for ILoggingError */
-export const isILoggingError = (x: any): x is ILoggingError => typeof x?.getTelemetryProperties === "function";
+/**
+ * Annotate the given error object, such that it becomes a valid Fluid Error (and is asserted so via a type guard)
+ * If the object is not suitable for annotation (e.g. it's frozen), this will throw.
+ * @param error - The error to annotate by patching in the properties of IFluidErrorBase
+ * @param annotations - Annotations to apply to the normalized error
+ */
+export function annotateErrorObject(
+    errorObject: unknown,
+    annotations: FluidErrorAnnotations = {},
+): asserts errorObject is IFluidErrorBase {
+    // If errorObject isn't suitable to annotate as IFluidErrorBase, bail.
+    assert(isRegularObject(errorObject) && !Object.isFrozen(errorObject),
```

Comment:
we should never assert in a place like this. we'll lose the original error, we need to be much more defensive here